webfrontend/src/components/navbar/search-user-box.tsx

```diff
@@ -0,0 +1,120 @@
+            fetch('/user/search_username?username=' + username.slice(1))
+                .then(response => {
+                    setIsSearching(false);
+                    if (response.status !== 200) {
+                        triggerErrorMessage();
+                    } else {
+                        response.json().then(response => setResults(response.users))
+                    }
+                });
+        } else {
+            fetch('/user/search?username=' + username)
+                .then(response => {
+                    setIsSearching(false);
+                    if (response.status !== 200) {
+                        triggerErrorMessage();
+                    } else {
+                        
+                        response.json().then(response => {
+                          setResults(response.users) 
+                        })
+                    }
+                });
+        }
+    }
+  }, [isSearching, username]);
+
+  const submitSearchField = useCallback((event: React.SyntheticEvent<EventTarget>) => {
+    event.preventDefault();
+    searchUser();
+  }, [searchUser])
+
+  const handleChange = useCallback((value: string) => {
+    setUserName(value);
+    setIsSearching(false);
+    searchUser();
+  }, [searchUser])
+ 
+  useEffect(() => {
+    window.onclick = (event: MouseEvent) => {
+      if (searchForm.current && searchForm.current!.contains(event.target as Node)) {
+        setIsClickOutside(false);
+      } else {
+        setIsClickOutside(true);
+      }
+    }
+ 
+  }, [clickOutside]);
+
+  const popover = (
+    <Popover id={`popover-positioned-bottom`}>
+    <Popover.Body>
+     {username !== '' ? results.length !== 0 ?
+        <div ref = {searchResultBox} id="search-result-box" style={{display: 'flex'}}>
+          <div className="result-lists">
+            {results.map((result: userProperties, id) => {
+              return (
+              <Link key={id}  to={`${result.profile_link}`} style={{textDecoration: 'none'}}>
+                <div className="result-item"> 
+                  <img style={{height: '25px'}} src={result.profile_pic_url} alt="" />
+                  <div className="user-info">
+                    <p className="user-name"> {result.name}</p>
+                    <p className="user-username">{result.username}</p>
+                  </div>
+                </div>
+              </Link>
+              )
+            })}
+          </div>
+        </div>: 
+          <div ref = {searchResultBox} id="search-result-box" style={{display: 'flex'}}>
+            <div id="search-box-no-result" className="text-body-tertiary">No user match your search</div>
+          </div> : 
+           <div id="search-box-placeholder" className='text-body-tertiary'>
+            <p> <strong>Type and search for user ...</strong></p>
+            <p> <strong>Tip: add '@' in front to search by username</strong></p>
+          </div>
+      }   
+    </Popover.Body>
+  </Popover>
+  );
```

Comment:
Please do not declare a JSX element as a constant inside another element. Rendering is very expensive. Constants in a component are redefined on every re-render, and hence this chunk of html is re-rendered on every change of state. Please declare a new component **outside of this component**, or declare a `useMemo` hook, or put this chunk of code directly to the places that it is used (should not do this if this chunk of code is used more than once).